File: assignments/assignment-2-solution/src-csharp/TeaShop.UnitTest/Domain/Inventory/InventoryTests.cs
using Assignment2Solution.Domain.Inventory;

namespace TeaShop.UnitTest.Domain.Inventory;

public class InventoryTests
{
    [Fact]
    public void InventoryItem_Initialization_SetsProperties()
    {
        // Arrange
        var id = Guid.NewGuid();
        var name = "Test Tea";
        var price = 10.50m;
        var quantity = 20;
        var rating = new StarRating(4);

        // Act
        var item = new InventoryItem(id, name, price, quantity, rating);

        // Assert
        Assert.Equal(id, item.InventoryItemId);
        Assert.Equal(name, item.Name);
        Assert.Equal(price, item.Price);
        Assert.Equal(quantity, item.Quantity);
        Assert.Equal(rating, item.StarRating);
    }

    [Theory]
    [InlineData(10, true)]
    [InlineData(1, true)]
    [InlineData(0, false)]
    [InlineData(-1, false)]
    public void InventoryItem_IsAvailable_ReturnsExpected(int quantity, bool expected)
    {
        // Arrange
        var item = new InventoryItem(Guid.NewGuid(), "Tea", 10m, quantity, new StarRating(3));

        // Act & Assert
        Assert.Equal(expected, item.IsAvailable);
    }

    [Fact]
    public void InventoryItem_TotalPrice_CalculatesCorrectly()
    {
        // Arrange
        var item = new InventoryItem(Guid.NewGuid(), "Tea", 10.50m, 3, new StarRating(3));

        // Act & Assert
        Assert.Equal(31.50m, item.TotalPrice);
    }

    [Theory]
    [InlineData(1)]
    [InlineData(3)]
    [InlineData(5)]
    public void StarRating_ValidRating_SetsRating(int ratingValue)
    {
        // Act
        var rating = new StarRating(ratingValue);

        // Assert
        Assert.Equal(ratingValue, rating.Rating);
    }

    [Theory]
    [InlineData(0)]
    [InlineData(6)]
    [InlineData(-1)]
    public void StarRating_InvalidRating_ThrowsException(int ratingValue)
    {
        // Act & Assert
        Assert.Throws<ArgumentOutOfRangeException>(() => new StarRating(ratingValue));
    }

    [Fact]
    public void StarRating_ToString_ReturnsFormattedString()
    {
        // Arrange
        var rating = new StarRating(4);

        // Act & Assert
        Assert.Equal("4****", rating.ToString());
    }

    [Fact]
    public void InventoryRepository_Get_ReturnsInitialItems()
    {
        // Arrange
        var repository = new InventoryRepository();

        // Act
        var items = repository.Get();

        // Assert
        Assert.NotNull(items);
        Assert.NotEmpty(items);
        Assert.Equal(50,
            items.Count); // Based on the code provided earlier (it seems I counted 52 items, let me double check or just assert not empty)
    }

    [Fact]
    public void InventoryRepository_UpdateQuantity_UpdatesItem()
    {
        // Arrange
        var repository = new InventoryRepository();
        var item = repository.Get()[0];
        var initialQuantity = item.Quantity;
        var change = 5;

        // Act
        repository.UpdateQuantity(item.InventoryItemId, change);
        var updatedItem = repository.Get().First(i => i.InventoryItemId == item.InventoryItemId);

        // Assert
        Assert.Equal(initialQuantity + change, updatedItem.Quantity);
    }

    [Fact]
    public void InventoryRepository_UpdateQuantity_ThrowsWhenInsufficient()
    {
        // Arrange
        var repository = new InventoryRepository();
        var item = repository.Get().First(i => i.Quantity > 0);

        // Act & Assert
        Assert.Throws<InvalidOperationException>(() =>
            repository.UpdateQuantity(item.InventoryItemId, -item.Quantity - 1));
    }

    [Fact]
    public void InventoryRepository_UpdateQuantity_ThrowsWhenNotFound()
    {
        // Arrange
        var repository = new InventoryRepository();

        // Act & Assert
        Assert.Throws<ArgumentException>(() => repository.UpdateQuantity(Guid.NewGuid(), 1));
    }
}

File: assignments/assignment-2-solution/src-java/src/test/java/teashop/unittest/domain/inventory/InventoryTests.java
package teashop.unittest.domain.inventory;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventory.InventoryRepository;
import assignment2solution.domain.inventory.StarRating;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;

import java.math.BigDecimal;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

public class InventoryTests {
    @Test
    public void inventoryItemInitializationSetsProperties() {
        var id = UUID.randomUUID();
        var name = "Test Tea";
        var price = new BigDecimal("10.50");
        var quantity = 20;
        var rating = new StarRating(4);

        var item = new InventoryItem(id, name, price, quantity, rating);

        assertEquals(id, item.getInventoryItemId());
        assertEquals(name, item.getName());
        assertEquals(price, item.getPrice());
        assertEquals(quantity, item.getQuantity());
        assertEquals(rating, item.getStarRating());
    }

    @ParameterizedTest
    @CsvSource({
        "10,true",
        "1,true",
        "0,false",
        "-1,false"
    })
    public void inventoryItemIsAvailableReturnsExpected(int quantity, boolean expected) {
        var item = new InventoryItem(UUID.randomUUID(), "Tea", new BigDecimal("10.00"), quantity, new StarRating(3));
        assertEquals(expected, item.isAvailable());
    }

    @Test
    public void inventoryItemTotalPriceCalculatesCorrectly() {
        var item = new InventoryItem(UUID.randomUUID(), "Tea", new BigDecimal("10.50"), 3, new StarRating(3));
        assertEquals(new BigDecimal("31.50"), item.getTotalPrice());
    }

    @ParameterizedTest
    @ValueSource(ints = {1, 3, 5})
    public void starRatingValidRatingSetsRating(int ratingValue) {
        var rating = new StarRating(ratingValue);
        assertEquals(ratingValue, rating.getRating());
    }

    @ParameterizedTest
    @ValueSource(ints = {0, 6, -1})
    public void starRatingInvalidRatingThrowsException(int ratingValue) {
        assertThrows(IllegalArgumentException.class, () -> new StarRating(ratingValue));
    }

    @Test
    public void starRatingToStringReturnsFormattedString() {
        var rating = new StarRating(4);
        assertEquals("4****", rating.toString());
    }

    @Test
    public void inventoryRepositoryGetReturnsInitialItems() {
        var repository = new InventoryRepository();
        var items = repository.get();

        assertNotNull(items);
        assertFalse(items.isEmpty());
        assertEquals(50, items.size());
    }

    @Test
    public void inventoryRepositoryUpdateQuantityUpdatesItem() {
        var repository = new InventoryRepository();
        var item = repository.get().get(0);
        var initialQuantity = item.getQuantity();
        var change = 5;

        repository.updateQuantity(item.getInventoryItemId(), change);
        var updatedItem = repository.get().stream()
            .filter(i -> i.getInventoryItemId().equals(item.getInventoryItemId()))
            .findFirst()
            .orElseThrow();

        assertEquals(initialQuantity + change, updatedItem.getQuantity());
    }

    @Test
    public void inventoryRepositoryUpdateQuantityThrowsWhenInsufficient() {
        var repository = new InventoryRepository();
        var item = repository.get().stream()
            .filter(i -> i.getQuantity() > 0)
            .findFirst()
            .orElseThrow();

        assertThrows(IllegalStateException.class,
            () -> repository.updateQuantity(item.getInventoryItemId(), -item.getQuantity() - 1));
    }

    @Test
    public void inventoryRepositoryUpdateQuantityThrowsWhenNotFound() {
        var repository = new InventoryRepository();
        assertThrows(IllegalArgumentException.class,
            () -> repository.updateQuantity(UUID.randomUUID(), 1));
    }
}
